opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp

```diff
@@ -130,10 +157,16 @@ class OwningTwoLevelPreconditioner : public Dune::PreconditionerWithUpdate<Vecto
         twolevel_method_.post(x);
     }
 
-    virtual void update() override
+    virtual void update(const VectorType& weights, const boost::property_tree::ptree& /*prm*/) override
     {
-        Opm::Amg::getQuasiImpesWeights<MatrixType, VectorType>(
-            linear_operator_.getmat(), prm_.get<int>("pressure_var_index"), transpose, weights_);
+	
+	// if(prm.get<std::string>("weight_type") == "quasiimpes"){
+	//     Opm::Amg::getQuasiImpesWeights<MatrixType, VectorType>(
+	// 	linear_operator_.getmat(), prm_.get<int>("pressure_var_index"), transpose, weights_);
+	//     weights_ = prm.get<VectorType>("weights");
+	// }else{
+	    weights_ = weights;
```

Comment:
Can someone please tell me, why this change make sense?

If I am not mistaking, we assume that even if the matrix is changing so much, that there definitely is a need to recalculate the preconditioner, we, nevertheless assume that the absolute maximum entry of x, solving e.g. `a_{ii} x = [ 0 1 0]^T` for any row `i` stays the same.

Previously, we always recalculated the weights. That seemed to make more sense.

What am I missing?

Reply:
Ok. I misinterpreted this. The weights do get recalculate, but this is a bit obfuscated as it is moved to ISTLSolverEbosFlexible. I am not smart enough for this distribution of code.

Now the question is why does this code moval makes sense,

Reply:
The reason might be that calculating the true impes weights needs the grid and information from opm-models and that should not be exposed to the constructors.

I do not agree 100%  here because the update method now gets passed information that is specific to some preconditioners. Will need to think, whether there is a better way.

Reply:
It some time ago I did this but. To make the liner solver independent of simulators. We need to pass in the new weights alternatively using store a reference to the vector and update it from the simulator. Quasimpes on the other hand can recalculate it from the matrix alone. I changed the update to have  both the weights and the parameters as input at some point to make it possible to do more advanced thing in the update. For example update in a particular way given information from the simulator (of coarse the full strategy for this could be in parameter tree used for the construction). The bad thing is that the update now has an interface with weights which specific for one linear solver.